src/Optimizer/Transformer/PreloadHeroImage.php

```diff
@@ -220,29 +237,33 @@ private function findHeroImages(Document $document)
             }
         }
 
-        // Optimize data-hero element if defined.
         if (count($heroImages) > 0) {
             return $heroImages;
         }
 
-        // Fall back to auto-detected hero image if available.
-        if ($heroImageCandidate) {
-            return [$heroImageCandidate];
+        while (count($heroImages) < self::DATA_HERO_MAX && count($heroImageCandidates) > 0) {
+            $heroImages[] = array_shift($heroImageCandidates);
+        }
+
+        if (count($heroImages) < 1 && count($heroImageFallbacks) > 0) {
```

Comment:
If the hero image candidates don't make up the maximum hero images allowed, should not the first 1-2 fallback images also backfill the hero images?

```suggestion
        while (count($heroImages) < self::DATA_HERO_MAX && count($heroImageFallbacks) > 0) {
```

Reply:
I had this in the beginning but then decided against it, as it can lead to unexpected problems.

The fallback images are basically collected from the first big images based on markup order. If you use something like we have with `DetermineHeroImages` to be smarter about what to pick, filling it up with a fallback image could mean the fallback image was something specifically skipped in `DetermineHeroImages` but still earlier in markup order than the actual candidate.

I'd prefer "fallback" images to be a fallback in case nothing at all had been marked, and as soon as _something_ was marked, we assume the marking was correct and no fallback are needed or will make things worse.